extensions-contrib/first-last-string/src/main/java/io/druid/query/aggregation/first/StringFirstAggregatorFactory.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.aggregation.first;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.primitives.Longs;
+import io.druid.collections.SerializablePair;
+import io.druid.java.util.common.StringUtils;
+import io.druid.query.aggregation.AggregateCombiner;
+import io.druid.query.aggregation.Aggregator;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.AggregatorUtil;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import io.druid.segment.BaseObjectColumnValueSelector;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.column.Column;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class StringFirstAggregatorFactory extends AggregatorFactory
+{
+  public static final Integer MAX_SIZE_STRING = 1024;
+  public static final Comparator VALUE_COMPARATOR = (o1, o2) -> ((SerializablePair<Long, String>) o1).rhs.equals(((SerializablePair<Long, String>) o2).rhs)
+                                                                ? 1
+                                                                : 0;
+
+  public static final Comparator TIME_COMPARATOR = (o1, o2) -> Longs.compare(
+      ((SerializablePair<Long, Object>) o1).lhs,
+      ((SerializablePair<Long, Object>) o2).lhs
+  );
+
+  private final String fieldName;
+  private final String name;
+  private final Integer maxStringBytes;
+
+  @JsonCreator
+  public StringFirstAggregatorFactory(
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") final String fieldName,
```

Comment:
It'd be cool if this (and last) supported expressions like the *sum and min/max aggregators

Reply:
Afraid there's not comprehensive documentation at the moment. I haven't been digging into Druid's source for too long so feel free to ignore me!

There's some documentation on the language itself here: https://github.com/druid-io/druid/blob/master/docs/content/misc/math-expr.md

I think adding support for it should be making the constructor of this class similar to LongsumAggregatorFactory, taking an injected macroTable and a string for the expression. 

Then in your factorize methods you'd check for an expression, if that's present you'd return an ExprEvalSelector instead of the results of the metricFactory.makeColumnValueSelector.

Making the ExprEvalSelector would look something like this:

`
final Expr expr = Parser.parse(fieldExpression, macroTable);
return ExpressionSelectors.makeExprEvalSelector(metricFactory, expr);
`

Reply:
Druid's expression system is not documented yet because it was an experimental feature (but maybe it's time to document). 

I think we can add `expression` field to here later.